### PR TITLE
805 create prisoner search api client

### PR DESCRIPTION
**Pull request by anandelangovan1**
## Context
https://trello.com/c/LKPG1BwN/805-integrate-with-the-prisoner-search-api-m

## Changes in this PR

As part of this PR we are adding a prison-search-api client with 1 methods to getPrisoners details by prisonId 
Setting up of the prison-search-api locally has been tricky and local testing is challenged by the lack of data. We have tested the endpoint by point at the dev env and we are able to get a response

## Release checklist

[Release process documentation](../doc/how-to/perform-a-release.md)

As part of our continuous deployment strategy we must ensure that this work is
ready to be released once merged.

### Pre-merge

- [ ] There are changes required to the Accredited Programmes UI for this change to work...
  - [ ] ... and they been released to production already

### Post-merge

- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-preprod-environment) release to preprod
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-production-environment) release to prod

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/config/WebClientConfiguration.kt [hunk 1]
```diff
@@ -74,7 +74,7 @@ class WebClientConfiguration(
   ): WebClient {
     val oauth2Client = ServletOAuth2AuthorizedClientExchangeFilterFunction(authorizedClientManager)
 
-    oauth2Client.setDefaultClientRegistrationId("prisonerSearchApiWebClient")
+    oauth2Client.setDefaultClientRegistrationId("prisoner-search-api")
```
Comment (Gweaton) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/config/WebClientConfiguration.kt:77]: Can we edit the commit before this one to include this change so we're not introducing an incorrect bit of code and then fixing it in the next commit? Should be able to do this with a `git rebase` and then editing the previous commit to include this one.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/client/PrisonerSearchApiClient.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package uk.gov.justice.digital.hmpps.hmppsaccreditedprogrammesapi.client
```
Comment (Gweaton) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/client/PrisonerSearchApiClient.kt:1]: Let's move the adding of the client, prisoner numbers and prisoners into their own commit, as this is unrelated to running the service with wiremock.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/response/model/Prisoners.kt [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package uk.gov.justice.digital.hmpps.hmppsaccreditedprogrammesapi.response.model
+
+import java.time.LocalDate
+
+
+data class Prisoner(
```
Comment (Gweaton) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/response/model/Prisoners.kt]: Is it worth putting this in a directory called `PrisonerSearch` like they do on the Approved Premises project [here](https://github.com/ministryofjustice/hmpps-approved-premises-api/tree/main/src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/model/prisonsapi). So we'd have `/hmppsaccreditedprogrammesapi/model/prisonerSearch/Prisoner.kt` I believe it should be singular too?

This would be a useful pattern for separating models in our own service from those from other APIs and grouping response types clearly.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/pact/PactContractTest.kt [hunk 4]
```diff
@@ -77,6 +77,9 @@ class PactContractTest {
   @State("Referral can be submitted")
   fun `ensure referral can be submitted`() {}
 
+  @State("Referral summaries exist for an organisation with the ID a026cc07-8e0b-40dd-9b66-1b3dacecc63d")
+  fun `ensure referral summaries exist for an organisation with the ID a026cc07-8e0b-40dd-9b66-1b3dacecc63d`() {}
```
Comment (nsmj63-dxw) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/pact/PactContractTest.kt:81]: I've also implemented this fix in my PR. Given it's to do with referral summaries it's probably best to axe this change in this PR; I'll merge mine in first so you can rebase.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/request/model/PrisonerNumbers.kt [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package uk.gov.justice.digital.hmpps.hmppsaccreditedprogrammesapi.request.model
```
Comment (nsmj63-dxw) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/request/model/PrisonerNumbers.kt:1]: I'm not sure this is the right place to put these requests. They're essentially db projections, so I think they should go into somewhere like `uk.gov.justice.digital.hmpps.hmppsaccreditedprogrammesapi.domain.entity.read`, which is where my projection goes for our internal data in my PR. Not precious about it, so am happy to talk through where best the package containing each of these should go.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/config/WebClientConfiguration.kt (hunk 1 above)
Comment (nsmj63-dxw) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/config/WebClientConfiguration.kt:77]: Agreed.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/client/PrisonerSearchApiClient.kt (hunk 2 above)
Comment (nsmj63-dxw) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsaccreditedprogrammesapi/client/PrisonerSearchApiClient.kt:1]: Agreed.

**Review by nsmj63-dxw — CHANGES_REQUESTED** [review]

**Review by Gweaton — APPROVED** [review]
